### PR TITLE
Workload API Sources Fixes and Tests

### PR DESCRIPTION
**Pull request by azdagron**
- Fixes bundle merging behavior in the BundleSource such that a lack of X.509 or JWT material for a given trust domain causes the relative bundle getters for X.509 or JWT bundles to return no bundle instead of an empty bundle
- Reduces code duplication code between the sources to simplify implementation.
- Adds WaitUntilUpdated methods to each source to allow callers to wait for changes. An Updated method that returns a channel that can be read from to detect changes is also added for callers that want to do their own select{}.
- Adds comprehensive unit-tests for the X509Source, JWTSource, and BundleSource.

### CODE REVIEW

File: v2/workloadapi/bundlesource.go [hunk 1]
```diff
@@ -98,31 +133,22 @@ func (s *BundleSource) setX509Context(x509Context *X509Context) {
...
 	trustDomains := make(map[spiffeid.TrustDomain]struct{}, len(newBundles))
 	for _, newBundle := range newBundles {
 		trustDomains[newBundle.TrustDomain()] = struct{}{}
-		existingBundle, ok := s.bundles.Get(newBundle.TrustDomain())
-		if !ok {
-			s.bundles.Add(spiffebundle.FromX509Bundle(newBundle))
-			continue
-		}
-		existingBundle.SetX509Authorities(newBundle.X509Authorities())
+		s.x509Authorities[newBundle.TrustDomain()] = newBundle.X509Authorities()
```
Comment (amartinezfayo) [inline v2/workloadapi/bundlesource.go:142]: This should be protected by a lock.

File: v2/workloadapi/bundlesource.go [hunk 2]
```diff
@@ -98,31 +133,22 @@ func (s *BundleSource) setX509Context(x509Context *X509Context) {
...
+	// represented in the X.509 context.
+	for existingTD := range s.x509Authorities {
+		if _, ok := trustDomains[existingTD]; ok {
 			continue
 		}
-		existingBundle.SetX509Authorities(nil)
-		if existingBundle.Empty() {
-			s.bundles.Remove(existingBundle.TrustDomain())
-		}
+		delete(s.x509Authorities, existingTD)
```
Comment (amartinezfayo) [inline v2/workloadapi/bundlesource.go:151]: This should be protected by a lock.

File: v2/workloadapi/bundlesource.go [hunk 3]
```diff
@@ -132,31 +158,22 @@ func (s *BundleSource) setJWTBundles(bundles *jwtbundle.Set) {
...
 	trustDomains := make(map[spiffeid.TrustDomain]struct{}, len(newBundles))
 	for _, newBundle := range newBundles {
 		trustDomains[newBundle.TrustDomain()] = struct{}{}
-		existingBundle, ok := s.bundles.Get(newBundle.TrustDomain())
-		if !ok {
-			s.bundles.Add(spiffebundle.FromJWTBundle(newBundle))
-			continue
-		}
-		existingBundle.SetJWTAuthorities(newBundle.JWTAuthorities())
+		s.jwtAuthorities[newBundle.TrustDomain()] = newBundle.JWTAuthorities()
```
Comment (amartinezfayo) [inline v2/workloadapi/bundlesource.go:167]: This should be protected by a lock.

File: v2/workloadapi/bundlesource.go [hunk 4]
```diff
@@ -132,31 +158,22 @@ func (s *BundleSource) setJWTBundles(bundles *jwtbundle.Set) {
...
+	// in the JWT bundles.
+	for existingTD := range s.jwtAuthorities {
+		if _, ok := trustDomains[existingTD]; ok {
 			continue
 		}
-		existingBundle.SetJWTAuthorities(nil)
-		if existingBundle.Empty() {
-			s.bundles.Remove(existingBundle.TrustDomain())
-		}
+		delete(s.jwtAuthorities, existingTD)
```
Comment (amartinezfayo) [inline v2/workloadapi/bundlesource.go:176]: This should be protected by a lock.

File: v2/workloadapi/bundlesource.go (hunk 1 above)
Comment (azdagron) [reply v2/workloadapi/bundlesource.go:142]: It is. See the top of the setX509Context function.

File: v2/workloadapi/bundlesource.go (hunk 4 above)
Comment (azdagron) [reply v2/workloadapi/bundlesource.go:176]: It is. See the top of the setJWTBundles function.

File: v2/workloadapi/bundlesource.go (hunk 3 above)
Comment (azdagron) [reply v2/workloadapi/bundlesource.go:167]: It is. See the top of the setJWTBundles function.

File: v2/workloadapi/bundlesource.go (hunk 2 above)
Comment (azdagron) [reply v2/workloadapi/bundlesource.go:151]: It is. See the top of the setX509Context function.

File: v2/workloadapi/bundlesource.go (hunk 1 above)
Comment (amartinezfayo) [reply v2/workloadapi/bundlesource.go:142]: Sorry, I didn't expand those lines in the GitHub UI 🤦

**Pull request merged**